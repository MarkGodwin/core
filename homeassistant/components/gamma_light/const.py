"""Constants for the Switch as X integration."""

from typing import Final

DOMAIN: Final = "gamma_light"
MIN_BRIGHTNESS: Final = "min_brightness"
GAMMA: Final = "gamma"
