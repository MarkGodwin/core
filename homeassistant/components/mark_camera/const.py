"""Constants for the generic (IP Camera) integration."""

DOMAIN = "mark_camera"
DEFAULT_NAME = "Mark's Generic Camera"
CONF_CONFIRMED_OK = "confirmed_ok"
CONF_CONTENT_TYPE = "content_type"
CONF_LIMIT_REFETCH_TO_URL_CHANGE = "limit_refetch_to_url_change"
CONF_STILL_IMAGE_URL = "still_image_url"
CONF_STREAM_SOURCE = "stream_source"
CONF_FRAMERATE = "framerate"
CONF_MOTION_SENSOR = "motion_sensor_id"
CONF_DOORBELL_SENSOR = "doorbell_sensor_id"
GET_IMAGE_TIMEOUT = 10

DEFAULT_USERNAME = None
DEFAULT_PASSWORD = None
DEFAULT_IMAGE_URL = None
DEFAULT_STREAM_SOURCE = None
