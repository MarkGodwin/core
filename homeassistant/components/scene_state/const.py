"""Constants for the Scene State integration."""

DOMAIN = "scene_state"
