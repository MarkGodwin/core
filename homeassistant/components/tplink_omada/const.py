"""Constants for the TP-Link Omada integration."""

DOMAIN = "tplink_omada"

CONF_SITE = "site"

OPT_DEVICE_TRACKER = "device_tracker"
OPT_TRACKED_CLIENTS = "tracked_clients"
OPT_SCANNED_CLIENTS = "scanned_clients"
OPT_TRACKER_POLL_INTERVAL = "tracker_poll_interval"

DEFAULT_TRACKER_POLL_INTERVAL = 300
