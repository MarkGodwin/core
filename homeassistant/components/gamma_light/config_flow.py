"""Config flow for gamma_light."""

from homeassistant.core import HomeAssistant
from homeassistant.const import Platform
from homeassistant.helpers import config_entry_flow
from homeassistant.helpers import entity_registry

from .const import DOMAIN


async def _async_has_devices(hass: HomeAssistant) -> bool:
    """Return if there are devices that can be discovered."""
    # TODO Check if there are any devices that can be discovered in the network.

    registry = entity_registry.async_get(hass)

    potential_lights = [
        entry
        for entry in registry.entities.values()
        if entry.domain == Platform.LIGHT and entry.platform != DOMAIN
    ]

    # Exclude any lights we have already wrapped

    return len(potential_lights) > 0
    # devices = await hass.async_add_executor_job(my_pypi_dependency.discover)
    # return len(devices) > 0


config_entry_flow.register_discovery_flow(DOMAIN, "gamma_light", _async_has_devices)
