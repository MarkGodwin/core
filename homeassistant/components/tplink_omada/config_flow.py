"""Config flow for TP-Link Omada integration."""

from __future__ import annotations

from collections.abc import Mapping
import logging
import re
from types import MappingProxyType
from typing import Any, NamedTuple
from urllib.parse import urlsplit

from aiohttp import CookieJar
from tplink_omada_client import OmadaClient, OmadaSite
from tplink_omada_client.clients import OmadaWirelessClient
from tplink_omada_client.exceptions import (
    ConnectionFailed,
    LoginFailed,
    OmadaClientException,
    UnsupportedControllerVersion,
)
import voluptuous as vol

from homeassistant.config_entries import (
    ConfigEntry,
    ConfigFlow,
    ConfigFlowResult,
    OptionsFlow,
    OptionsFlowWithConfigEntry,
)
from homeassistant.const import (
    CONF_HOST,
    CONF_PASSWORD,
    CONF_USERNAME,
    CONF_VERIFY_SSL,
    UnitOfTime,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers import selector
from homeassistant.helpers.aiohttp_client import (
    async_create_clientsession,
    async_get_clientsession,
)

from .const import (
    CONF_SITE,
    DEFAULT_TRACKER_POLL_INTERVAL,
    DOMAIN,
    OPT_DEVICE_TRACKER,
    OPT_SCANNED_CLIENTS,
    OPT_TRACKED_CLIENTS,
    OPT_TRACKER_POLL_INTERVAL,
)
from .controller import OmadaSiteController

_LOGGER = logging.getLogger(__name__)

_STEP_DEVICE_TRACKER = "device_tracker"

STEP_USER_DATA_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_HOST): str,
        vol.Required(CONF_VERIFY_SSL, default=True): bool,
        vol.Required(CONF_USERNAME): str,
        vol.Required(CONF_PASSWORD): str,
    }
)


async def create_omada_client(
    hass: HomeAssistant, data: MappingProxyType[str, Any]
) -> OmadaClient:
    """Create a TP-Link Omada client API for the given config entry."""

    host: str = data[CONF_HOST]
    verify_ssl = bool(data[CONF_VERIFY_SSL])

    if not host.lower().startswith(("http://", "https://")):
        host = "https://" + host
    host_parts = urlsplit(host)
    if (
        host_parts.hostname
        and re.fullmatch(r"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}", host_parts.hostname)
        is not None
    ):
        # TP-Link API uses cookies for login session, so an unsafe cookie jar is required for IP addresses
        websession = async_create_clientsession(
            hass, cookie_jar=CookieJar(unsafe=True), verify_ssl=verify_ssl
        )
    else:
        websession = async_get_clientsession(hass, verify_ssl=verify_ssl)

    username = data[CONF_USERNAME]
    password = data[CONF_PASSWORD]

    return OmadaClient(host, username, password, websession=websession)


class HubInfo(NamedTuple):
    """Discovered controller information."""

    controller_id: str
    name: str
    sites: list[OmadaSite]


async def _validate_input(hass: HomeAssistant, data: dict[str, Any]) -> HubInfo:
    """Validate the user input allows us to connect."""

    client = await create_omada_client(hass, MappingProxyType(data))
    controller_id = await client.login()
    name = await client.get_controller_name()
    sites = await client.get_sites()

    return HubInfo(controller_id, name, sites)


class TpLinkOmadaConfigFlow(ConfigFlow, domain=DOMAIN):
    """Handle a config flow for TP-Link Omada."""

    VERSION = 1

    def __init__(self) -> None:
        """Create the config flow for a new integration."""
        self._omada_opts: dict[str, Any] = {}
        self._sites: list[OmadaSite] = []
        self._controller_name = ""

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the initial step."""

        errors: dict[str, str] = {}
        info = None
        if user_input is not None:
            info = await self._test_login(user_input, errors)

        if info is None or user_input is None:
            return self.async_show_form(
                step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
            )

        await self.async_set_unique_id(info.controller_id)
        self._abort_if_unique_id_configured()

        self._omada_opts.update(user_input)
        self._sites = info.sites
        self._controller_name = info.name
        if len(self._sites) > 1:
            return await self.async_step_site()
        return await self.async_step_site({CONF_SITE: self._sites[0].id})

    async def async_step_site(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle step to select site to manage."""

        if user_input is None:
            schema = vol.Schema(
                {
                    vol.Required(CONF_SITE, "site"): selector.SelectSelector(
                        selector.SelectSelectorConfig(
                            options=[
                                selector.SelectOptionDict(value=s.id, label=s.name)
                                for s in self._sites
                            ],
                            multiple=False,
                            mode=selector.SelectSelectorMode.DROPDOWN,
                        )
                    )
                }
            )

            return self.async_show_form(step_id="site", data_schema=schema)

        self._omada_opts.update(user_input)
        site_name = next(
            site for site in self._sites if site.id == user_input["site"]
        ).name
        display_name = f"{self._controller_name} ({site_name})"

        return self.async_create_entry(title=display_name, data=self._omada_opts)

    async def async_step_reauth(
        self, entry_data: Mapping[str, Any]
    ) -> ConfigFlowResult:
        """Perform reauth upon an API authentication error."""
        self._omada_opts = dict(entry_data)
        return await self.async_step_reauth_confirm()

    async def async_step_reauth_confirm(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Dialog that informs the user that reauth is required."""

        errors: dict[str, str] = {}

        if user_input is not None:
            self._omada_opts.update(user_input)
            info = await self._test_login(self._omada_opts, errors)

            if info is not None:
                # Auth successful - update the config entry with the new credentials
                entry = self.hass.config_entries.async_get_entry(
                    self.context["entry_id"]
                )
                assert entry is not None
                self.hass.config_entries.async_update_entry(
                    entry, data=self._omada_opts
                )
                await self.hass.config_entries.async_reload(entry.entry_id)
                return self.async_abort(reason="reauth_successful")

        return self.async_show_form(
            step_id="reauth_confirm",
            data_schema=vol.Schema(
                {
                    vol.Required(CONF_USERNAME): str,
                    vol.Required(CONF_PASSWORD): str,
                }
            ),
            errors=errors,
        )

    async def _test_login(
        self, data: dict[str, Any], errors: dict[str, str]
    ) -> HubInfo | None:
        try:
            info = await _validate_input(self.hass, data)
            if len(info.sites) > 0:
                return info
            errors["base"] = "no_sites_found"

        except ConnectionFailed:
            errors["base"] = "cannot_connect"
        except LoginFailed:
            errors["base"] = "invalid_auth"
        except UnsupportedControllerVersion:
            errors["base"] = "unsupported_controller"
        except OmadaClientException as ex:
            _LOGGER.error("Unexpected API error: %s", ex)
            errors["base"] = "unknown"
        except Exception:  # pylint: disable=broad-except
            _LOGGER.exception("Unexpected exception")
            errors["base"] = "unknown"
        return None

    @staticmethod
    @callback
    def async_get_options_flow(
        config_entry: ConfigEntry,
    ) -> OptionsFlow:
        """Create the options flow."""
        return OptionsFlowHandler(config_entry)


class OptionsFlowHandler(OptionsFlowWithConfigEntry):
    """Options flow for TP-Link Omada."""

    async def async_step_init(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Manage the top-level options."""

        self.options.update(user_input or {})
        device_tracker_enabled = self.options.get(OPT_DEVICE_TRACKER, False)

        if user_input is None:
            return self.async_show_form(
                step_id="init",
                data_schema=vol.Schema(
                    {
                        vol.Required(
                            OPT_DEVICE_TRACKER,
                            default=device_tracker_enabled,
                        ): bool
                    }
                ),
            )

        if device_tracker_enabled:
            return await self.async_step_device_tracker()

        return self.async_create_entry(data=self.options)

    async def async_step_device_tracker(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Manage the device tracker options."""
        errors = {}

        if user_input is not None:
            self.options.update(user_input)
            if (
                len(self.options[OPT_TRACKED_CLIENTS]) == 0
                and len(self.options[OPT_SCANNED_CLIENTS]) == 0
            ):
                errors[OPT_TRACKED_CLIENTS] = "no_clients_selected"

        if user_input is None or errors:
            controller: OmadaSiteController = self.hass.data[DOMAIN][
                self.config_entry.entry_id
            ]

            site_client = controller.omada_client
            # Get the known Wi-Fi clients, in alphabetical order moving un-named clients to the end of the list
            clients: list[OmadaWirelessClient] = sorted(
                [
                    client
                    async for client in site_client.get_known_clients()
                    if isinstance(client, OmadaWirelessClient)
                ],
                key=lambda c: (0 if c.name != c.mac else 1, c.name),
            )

            schema = vol.Schema(
                {
                    vol.Optional(
                        OPT_SCANNED_CLIENTS,
                        default=self.options.get(OPT_SCANNED_CLIENTS, []),
                    ): _get_clients_selector(clients),
                    vol.Optional(
                        OPT_TRACKED_CLIENTS,
                        default=self.options.get(OPT_TRACKED_CLIENTS, []),
                    ): _get_clients_selector(clients),
                    vol.Required(
                        OPT_TRACKER_POLL_INTERVAL,
                        default=self.options.get(
                            OPT_TRACKER_POLL_INTERVAL, DEFAULT_TRACKER_POLL_INTERVAL
                        ),
                    ): selector.NumberSelector(
                        selector.NumberSelectorConfig(
                            min=30,
                            max=600,
                            step=5,
                            unit_of_measurement=UnitOfTime.SECONDS,
                            mode=selector.NumberSelectorMode.SLIDER,
                        )
                    ),
                }
            )

            return self.async_show_form(
                step_id=_STEP_DEVICE_TRACKER, data_schema=schema, errors=errors
            )

        return self.async_create_entry(data=self.options)


def _get_clients_selector(
    clients: list[OmadaWirelessClient],
) -> selector.SelectSelector:
    return selector.SelectSelector(
        selector.SelectSelectorConfig(
            options=[
                selector.SelectOptionDict(value=c.mac, label=c.name) for c in clients
            ],
            multiple=True,
            mode=selector.SelectSelectorMode.DROPDOWN,
        )
    )
