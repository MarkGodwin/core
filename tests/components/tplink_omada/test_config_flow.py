"""Test the TP-Link Omada config flows."""

from unittest.mock import MagicMock, patch

from syrupy.assertion import SnapshotAssertion
from tplink_omada_client import OmadaSite
from tplink_omada_client.exceptions import (
    ConnectionFailed,
    LoginFailed,
    OmadaClientException,
    UnsupportedControllerVersion,
)

from homeassistant import config_entries
from homeassistant.components.tplink_omada.config_flow import (
    OPT_DEVICE_TRACKER,
    OPT_SCANNED_CLIENTS,
    OPT_TRACKED_CLIENTS,
    HubInfo,
    _validate_input,
    create_omada_client,
)
from homeassistant.components.tplink_omada.const import DOMAIN
from homeassistant.core import HomeAssistant
from homeassistant.data_entry_flow import FlowResultType

from tests.common import MockConfigEntry

MOCK_USER_DATA = {
    "host": "https://fake.omada.host",
    "verify_ssl": True,
    "username": "test-username",
    "password": "test-password",
}

MOCK_ENTRY_DATA = {
    "host": "https://fake.omada.host",
    "verify_ssl": True,
    "site": "SiteId",
    "username": "test-username",
    "password": "test-password",
}


async def test_form_single_site(hass: HomeAssistant) -> None:
    """Test we get the form."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )
    assert result["type"] is FlowResultType.FORM
    assert result["errors"] == {}

    with (
        patch(
            "homeassistant.components.tplink_omada.config_flow._validate_input",
            return_value=HubInfo(
                "omada_id", "OC200", [OmadaSite("Display Name", "SiteId")]
            ),
        ) as mocked_validate,
        patch(
            "homeassistant.components.tplink_omada.async_setup_entry",
            return_value=True,
        ) as mock_setup_entry,
    ):
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            MOCK_USER_DATA,
        )
        await hass.async_block_till_done()

    assert result2["type"] is FlowResultType.CREATE_ENTRY
    assert result2["title"] == "OC200 (Display Name)"
    assert result2["data"] == MOCK_ENTRY_DATA
    assert len(mock_setup_entry.mock_calls) == 1
    mocked_validate.assert_called_once_with(hass, MOCK_USER_DATA)


async def test_form_multiple_sites(hass: HomeAssistant) -> None:
    """Test we get the form."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )
    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "user"
    assert result["errors"] == {}

    with (
        patch(
            "homeassistant.components.tplink_omada.config_flow._validate_input",
            return_value=HubInfo(
                "omada_id",
                "OC200",
                [OmadaSite("Site 1", "first"), OmadaSite("Site 2", "second")],
            ),
        ),
        patch(
            "homeassistant.components.tplink_omada.async_setup_entry",
            return_value=True,
        ) as mock_setup_entry,
    ):
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            MOCK_USER_DATA,
        )
        await hass.async_block_till_done()

    assert result2["type"] is FlowResultType.FORM
    assert result2["step_id"] == "site"

    with patch(
        "homeassistant.components.tplink_omada.async_setup_entry",
        return_value=True,
    ) as mock_setup_entry:
        result3 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                "site": "second",
            },
        )
        await hass.async_block_till_done()

    assert result3["type"] is FlowResultType.CREATE_ENTRY
    assert result3["title"] == "OC200 (Site 2)"
    assert result3["data"] == {
        "host": "https://fake.omada.host",
        "verify_ssl": True,
        "site": "second",
        "username": "test-username",
        "password": "test-password",
    }
    assert len(mock_setup_entry.mock_calls) == 1


async def test_form_invalid_auth(hass: HomeAssistant) -> None:
    """Test we handle invalid auth."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "homeassistant.components.tplink_omada.config_flow._validate_input",
        side_effect=LoginFailed(-1000, "Invalid username/password"),
    ):
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            MOCK_USER_DATA,
        )

    assert result2["type"] is FlowResultType.FORM
    assert result2["errors"] == {"base": "invalid_auth"}


async def test_form_api_error(hass: HomeAssistant) -> None:
    """Test we handle unknown API error."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "homeassistant.components.tplink_omada.config_flow._validate_input",
        side_effect=OmadaClientException,
    ):
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            MOCK_USER_DATA,
        )

    assert result2["type"] is FlowResultType.FORM
    assert result2["errors"] == {"base": "unknown"}


async def test_form_generic_exception(hass: HomeAssistant) -> None:
    """Test we handle unknown API error."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "homeassistant.components.tplink_omada.config_flow._validate_input",
        side_effect=Exception,
    ):
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            MOCK_USER_DATA,
        )

    assert result2["type"] is FlowResultType.FORM
    assert result2["errors"] == {"base": "unknown"}


async def test_form_unsupported_controller(hass: HomeAssistant) -> None:
    """Test we handle unknown API error."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "homeassistant.components.tplink_omada.config_flow._validate_input",
        side_effect=UnsupportedControllerVersion("4.0.0"),
    ):
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            MOCK_USER_DATA,
        )

    assert result2["type"] is FlowResultType.FORM
    assert result2["errors"] == {"base": "unsupported_controller"}


async def test_form_cannot_connect(hass: HomeAssistant) -> None:
    """Test we handle cannot connect error."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "homeassistant.components.tplink_omada.config_flow._validate_input",
        side_effect=ConnectionFailed,
    ):
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            MOCK_USER_DATA,
        )

    assert result2["type"] is FlowResultType.FORM
    assert result2["errors"] == {"base": "cannot_connect"}


async def test_form_no_sites(hass: HomeAssistant) -> None:
    """Test we handle invalid auth."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "homeassistant.components.tplink_omada.config_flow._validate_input",
        return_value=HubInfo("omada_id", "OC200", []),
    ):
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            MOCK_USER_DATA,
        )

    assert result2["type"] is FlowResultType.FORM
    assert result2["errors"] == {"base": "no_sites_found"}


async def test_async_step_reauth_success(hass: HomeAssistant) -> None:
    """Test reauth starts an interactive flow."""

    mock_entry = MockConfigEntry(
        domain="tplink_omada",
        data=dict(MOCK_ENTRY_DATA),
        unique_id="USERID",
    )
    mock_entry.add_to_hass(hass)

    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={
            "source": config_entries.SOURCE_REAUTH,
            "entry_id": mock_entry.entry_id,
        },
        data=mock_entry.data,
    )

    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "reauth_confirm"

    with patch(
        "homeassistant.components.tplink_omada.config_flow._validate_input",
        return_value=HubInfo(
            "omada_id", "OC200", [OmadaSite("Display Name", "SiteId")]
        ),
    ) as mocked_validate:
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"], {"username": "new_uname", "password": "new_passwd"}
        )
    await hass.async_block_till_done()

    assert result2["type"] is FlowResultType.ABORT
    assert result2["reason"] == "reauth_successful"
    mocked_validate.assert_called_once_with(
        hass,
        {
            "host": "https://fake.omada.host",
            "verify_ssl": True,
            "site": "SiteId",
            "username": "new_uname",
            "password": "new_passwd",
        },
    )


async def test_async_step_reauth_invalid_auth(hass: HomeAssistant) -> None:
    """Test reauth starts an interactive flow."""

    mock_entry = MockConfigEntry(
        domain="tplink_omada",
        data=dict(MOCK_ENTRY_DATA),
        unique_id="USERID",
    )
    mock_entry.add_to_hass(hass)

    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={
            "source": config_entries.SOURCE_REAUTH,
            "entry_id": mock_entry.entry_id,
        },
        data=mock_entry.data,
    )

    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "reauth_confirm"

    with patch(
        "homeassistant.components.tplink_omada.config_flow._validate_input",
        side_effect=LoginFailed(-1000, "Invalid username/password"),
    ):
        result2 = await hass.config_entries.flow.async_configure(
            result["flow_id"], {"username": "new_uname", "password": "new_passwd"}
        )
    await hass.async_block_till_done()

    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "reauth_confirm"
    assert result2["errors"] == {"base": "invalid_auth"}


async def test_validate_input(hass: HomeAssistant) -> None:
    """Test validate returns HubInfo."""

    with (
        patch(
            "tplink_omada_client.omadaclient.OmadaClient", autospec=True
        ) as mock_client,
        patch(
            "homeassistant.components.tplink_omada.config_flow.create_omada_client",
            return_value=mock_client,
        ) as create_mock,
    ):
        mock_client.login.return_value = "Id"
        mock_client.get_controller_name.return_value = "Name"
        mock_client.get_sites.return_value = [OmadaSite("x", "y")]
        result = await _validate_input(hass, MOCK_USER_DATA)

    create_mock.assert_awaited_once()
    mock_client.login.assert_awaited_once()
    mock_client.get_controller_name.assert_awaited_once()
    mock_client.get_sites.assert_awaited_once()
    assert result.controller_id == "Id"
    assert result.name == "Name"
    assert result.sites == [OmadaSite("x", "y")]


async def test_create_omada_client_parses_args(hass: HomeAssistant) -> None:
    """Test config arguments are passed to Omada client."""

    with (
        patch(
            "homeassistant.components.tplink_omada.config_flow.OmadaClient",
            autospec=True,
        ) as mock_client,
        patch(
            "homeassistant.components.tplink_omada.config_flow.async_get_clientsession",
            return_value="ws",
        ) as mock_clientsession,
    ):
        result = await create_omada_client(hass, MOCK_USER_DATA)

    assert result is not None
    mock_client.assert_called_once_with(
        "https://fake.omada.host", "test-username", "test-password", "ws"
    )
    mock_clientsession.assert_called_once_with(hass, verify_ssl=True)


async def test_create_omada_client_adds_missing_scheme(hass: HomeAssistant) -> None:
    """Test config arguments are passed to Omada client."""

    with (
        patch(
            "homeassistant.components.tplink_omada.config_flow.OmadaClient",
            autospec=True,
        ) as mock_client,
        patch(
            "homeassistant.components.tplink_omada.config_flow.async_get_clientsession",
            return_value="ws",
        ) as mock_clientsession,
    ):
        result = await create_omada_client(
            hass,
            {
                "host": "fake.omada.host",
                "verify_ssl": True,
                "username": "test-username",
                "password": "test-password",
            },
        )

    assert result is not None
    mock_client.assert_called_once_with(
        "https://fake.omada.host", "test-username", "test-password", "ws"
    )
    mock_clientsession.assert_called_once_with(hass, verify_ssl=True)


async def test_create_omada_client_with_ip_creates_clientsession(
    hass: HomeAssistant,
) -> None:
    """Test config arguments are passed to Omada client."""

    with (
        patch(
            "homeassistant.components.tplink_omada.config_flow.OmadaClient",
            autospec=True,
        ) as mock_client,
        patch(
            "homeassistant.components.tplink_omada.config_flow.CookieJar", autospec=True
        ) as mock_jar,
        patch(
            "homeassistant.components.tplink_omada.config_flow.async_create_clientsession",
            return_value="ws",
        ) as mock_create_clientsession,
    ):
        result = await create_omada_client(
            hass,
            {
                "host": "10.10.10.10",
                "verify_ssl": True,
                "username": "test-username",
                "password": "test-password",
            },
        )

    assert result is not None
    mock_client.assert_called_once_with(
        "https://10.10.10.10", "test-username", "test-password", "ws"
    )
    mock_create_clientsession.assert_called_once_with(
        hass, cookie_jar=mock_jar.return_value, verify_ssl=True
    )


async def test_first_options_flow_with_no_tracker_creates_options(
    hass: HomeAssistant,
) -> None:
    """Test we get the initial form if there are no Options configured."""
    mock_entry = MockConfigEntry(
        domain="tplink_omada",
        data=dict(MOCK_ENTRY_DATA),
        unique_id="USERID",
    )
    mock_entry.add_to_hass(hass)

    result = await hass.config_entries.options.async_init(mock_entry.entry_id)

    assert result["type"] == FlowResultType.FORM
    assert result["step_id"] == "init"
    assert result["errors"] is None

    result = await hass.config_entries.options.async_configure(
        result["flow_id"], user_input={OPT_DEVICE_TRACKER: False}
    )
    assert result["type"] is FlowResultType.CREATE_ENTRY
    assert mock_entry.options == {OPT_DEVICE_TRACKER: False}


async def test_options_flow_defaults_form_with_selected_option(
    hass: HomeAssistant,
) -> None:
    """Test we get the initial form with pre-filled options."""
    mock_entry = MockConfigEntry(
        domain="tplink_omada",
        data=dict(MOCK_ENTRY_DATA),
        options={OPT_DEVICE_TRACKER: True},
        unique_id="USERID",
    )
    mock_entry.add_to_hass(hass)

    result = await hass.config_entries.options.async_init(mock_entry.entry_id)

    assert result["type"] == FlowResultType.FORM
    assert _get_schema_default(result["data_schema"].schema, OPT_DEVICE_TRACKER) is True
    assert result["step_id"] == "init"
    assert result["errors"] is None


def _get_schema_default(schema, key_name):
    for schema_key in schema:
        if schema_key == key_name:
            return schema_key.default()
    raise KeyError(f"{key_name} not found in schema")


async def test_options_flow_enable_device_tracker_gets_known_clients(
    hass: HomeAssistant,
    mock_omada_clients_only_client: MagicMock,
    mock_omada_clients_only_site_client: MagicMock,
    snapshot: SnapshotAssertion,
) -> None:
    """Test that if the device tracker is enabled, we get a populated device tracker form."""
    mock_entry = MockConfigEntry(
        domain="tplink_omada",
        data=dict(MOCK_ENTRY_DATA),
        options={OPT_DEVICE_TRACKER: False},
        unique_id="USERID",
    )
    mock_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(mock_entry.entry_id)
    await hass.async_block_till_done()

    result = await hass.config_entries.options.async_init(mock_entry.entry_id)
    assert result["type"] == FlowResultType.FORM
    result = await hass.config_entries.options.async_configure(
        result["flow_id"], user_input={OPT_DEVICE_TRACKER: True}
    )

    mock_omada_clients_only_site_client.get_known_clients.assert_called_once()
    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "device_tracker"
    schema = result["data_schema"].schema[OPT_TRACKED_CLIENTS]
    assert schema.config == snapshot


async def test_options_flow_enable_device_tracker_no_clients_selected_returns_error(
    hass: HomeAssistant,
    mock_omada_clients_only_client: MagicMock,
    mock_omada_clients_only_site_client: MagicMock,
) -> None:
    """Test that if the user doesn't select any devices to track, they are given an error message."""
    mock_entry = MockConfigEntry(
        domain="tplink_omada",
        data=dict(MOCK_ENTRY_DATA),
        options={OPT_DEVICE_TRACKER: False},
        unique_id="USERID",
    )
    mock_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(mock_entry.entry_id)
    await hass.async_block_till_done()

    result = await hass.config_entries.options.async_init(mock_entry.entry_id)
    assert result["type"] == FlowResultType.FORM
    result = await hass.config_entries.options.async_configure(
        result["flow_id"], user_input={OPT_DEVICE_TRACKER: True}
    )
    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "device_tracker"

    result = await hass.config_entries.options.async_configure(
        result["flow_id"],
        user_input={
            OPT_TRACKED_CLIENTS: [],
            OPT_SCANNED_CLIENTS: [],
        },
    )
    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "device_tracker"
    assert result["errors"] == {OPT_TRACKED_CLIENTS: "no_clients_selected"}


async def test_options_flow_enable_device_tracker_creates_entry(
    hass: HomeAssistant,
    mock_omada_clients_only_client: MagicMock,
    mock_omada_clients_only_site_client: MagicMock,
) -> None:
    """Test the options flow allows selection of tracked and scanned devices by Mac."""
    mock_entry = MockConfigEntry(
        domain="tplink_omada",
        data=dict(MOCK_ENTRY_DATA),
        options={OPT_DEVICE_TRACKER: False},
        unique_id="USERID",
    )
    mock_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(mock_entry.entry_id)
    await hass.async_block_till_done()

    result = await hass.config_entries.options.async_init(mock_entry.entry_id)
    assert result["type"] == FlowResultType.FORM
    result = await hass.config_entries.options.async_configure(
        result["flow_id"], user_input={OPT_DEVICE_TRACKER: True}
    )

    mock_omada_clients_only_site_client.get_known_clients.assert_called_once()
    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "device_tracker"

    result = await hass.config_entries.options.async_configure(
        result["flow_id"],
        user_input={
            OPT_TRACKED_CLIENTS: ["2E-DC-E1-C4-37-D3"],
            OPT_SCANNED_CLIENTS: ["2C-71-FF-ED-34-83"],
        },
    )
    assert result["type"] is FlowResultType.CREATE_ENTRY
    assert mock_entry.options == {
        OPT_DEVICE_TRACKER: True,
        OPT_TRACKED_CLIENTS: ["2E-DC-E1-C4-37-D3"],
        OPT_SCANNED_CLIENTS: ["2C-71-FF-ED-34-83"],
    }
