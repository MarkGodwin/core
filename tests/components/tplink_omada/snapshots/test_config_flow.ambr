# serializer version: 1
# name: test_options_flow_enable_device_tracker_gets_known_clients
  dict({
    'custom_value': False,
    'mode': 'dropdown',
    'multiple': True,
    'options': list([
      dict({
        'label': 'Apple',
        'value': '2E-DC-E1-C4-37-D3',
      }),
      dict({
        'label': 'Banana',
        'value': '2C-71-FF-ED-34-83',
      }),
      dict({
        'label': 'Pear',
        'value': '2C-D2-6B-BA-9C-94',
      }),
      dict({
        'label': '16-32-50-ED-FB-15',
        'value': '16-32-50-ED-FB-15',
      }),
    ]),
    'sort': False,
  })
# ---
